citest/json_predicate/path_predicate.py

```diff
@@ -0,0 +1,349 @@
+    return self.__path
+
+  @property
+  def pred(self):
+    """A delegate predicate.ValuePredicate used to find acceptable values."""
+    return self.__pred
+
+  def export_to_json_snapshot(self, snapshot, entity):
+    """Implements JsonSnapshotable interface."""
+    snapshot.edge_builder.make_control(entity, 'Path', self.__path)
+    snapshot.edge_builder.make_mechanism(entity, 'Predicate', self.__pred)
+
+  def __init__(self, path, pred=None):
+    """Construct finder instance.
+    Args:
+      path: The path to the field that we would like to find.
+      pred: The predicate.ValuePredicate to apply to the field we find.
+         This can be None indicating to just find the value at the field.
+    """
+    self.__pred = pred
+    self.__path = path or ''
+
+  def __eq__(self, finder):
+    return (self.__class__ == finder.__class__
+            and self.__path == finder.path
+            and self.__pred == finder.pred)
+
+  def __str__(self):
+    return '"{path}" {pred}'.format(path=self.__path, pred=self.__pred)
+
+  def __call__(self, source):
+    """Attempt to lookup the field in a JSON object.
+    Args:
+      source: JSON object to lookup within.
+
+    Returns:
+      PredicateResult on the bound predicate applied to the lookup path.
+        (i.e. pred(lookup(source, path)))
+    """
+
+    path = self.__path
+
+    builder = PathPredicateResultBuilder(pred=self, source=source)
+    enumerate_terminal = True
+    if path and path[-1] in (PATH_SEP, DONT_ENUMERATE_TERMINAL):
+      enumerate_terminal = path[-1] != DONT_ENUMERATE_TERMINAL
+      path = path[:-1]
+
+    queue = [_QueueElement(0, PathValue('', source))]
+    if not path and not (enumerate_terminal and isinstance(source, list)):
+      return self.__add_queue_to_builder(builder, queue, enumerate_terminal)
+
+    while queue:
+      if queue[0].path_offset >= len(path):
+        break
+
+      top = queue.pop(0)
+      candidates, fails = _process_queue_element(top, path)
+      queue.extend(candidates)
+      builder.add_all_path_failures(fails)
+
+    return self.__add_queue_to_builder(builder, queue, enumerate_terminal)
+
+  def __add_queue_to_builder(self, builder, queue, enumerate_terminal):
+    """Helper method for processing the final candidates from the queue.
+
+    Apply the filter bound to this predicate, if any, to determine whether
+    each of the final candidates should be kept or rejected.
+
+    Args:
+      builder: [PathPredicateResultBuilder] To add the results into.
+      queue: [list of _QueueElement] The final candidate values.
+      enumerate_terminal: [bool] If true, then list values in the queue
+         should be enumerated (one level) into individual elements.
+
+    Returns:
+      PathPredicateResult
+    """
+    for elem in queue:
+      value = elem.path_value.value
+      if enumerate_terminal and isinstance(value, list):
+        candidates, fails = _process_queue_element(elem, self.__path)
```

Comment:
We've already exhausted the path when we get into this function. All the elements in the queue roll up to the path. Otherwise we'd be doing more iterations in the outer function calling here.